File: apps/game-fps-monitor/desktop/src/main/adb.ts
import { exec, spawn } from 'child_process';
import { promisify } from 'util';

const execAsync = promisify(exec);

export interface AdbDevice {
  serial: string;
  state: string;
  model?: string;
}

export class AdbManager {
  private adbPath: string = 'adb';

  constructor(adbPath?: string) {
    if (adbPath) {
      this.adbPath = adbPath;
    }
  }

  async getDevices(): Promise<AdbDevice[]> {
    try {
      const { stdout } = await execAsync(`${this.adbPath} devices -l`);
      const lines = stdout.trim().split('\n').slice(1);
      
      return lines
        .filter((line) => line.trim())
        .map((line) => {
          const parts = line.split(/\s+/);
          const serial = parts[0];
          const state = parts[1];
          const modelMatch = line.match(/model:(\S+)/);
          
          return {
            serial,
            state,
            model: modelMatch ? modelMatch[1] : undefined,
          };
        })
        .filter((device) => device.state === 'device');
    } catch (error) {
      console.error('ADB设备检测失败:', error);
      return [];
    }
  }

  async forward(serial: string, localPort: number, remotePort: number): Promise<boolean> {
    try {
      await execAsync(
        `${this.adbPath} -s ${serial} forward tcp:${localPort} tcp:${remotePort}`
      );
      return true;
    } catch (error) {
      console.error('ADB端口转发失败:', error);
      return false;
    }
  }

  async removeForward(serial: string, localPort: number): Promise<boolean> {
    try {
      await execAsync(`${this.adbPath} -s ${serial} forward --remove tcp:${localPort}`);
      return true;
    } catch (error) {
      console.error('ADB端口转发移除失败:', error);
      return false;
    }
  }

  async getDeviceInfo(serial: string): Promise<Record<string, string>> {
    try {
      const { stdout } = await execAsync(
        `${this.adbPath} -s ${serial} shell getprop`
      );
      
      const info: Record<string, string> = {};
      const lines = stdout.split('\n');
      
      for (const line of lines) {
        const match = line.match(/\[(.+?)\]:\s*\[(.+?)\]/);
        if (match) {
          info[match[1]] = match[2];
        }
      }
      
      return info;
    } catch (error) {
      console.error('获取设备信息失败:', error);
      return {};
    }
  }

  async isAdbAvailable(): Promise<boolean> {
    try {
      await execAsync(`${this.adbPath} version`);
      return true;
    } catch {
      return false;
    }
  }
}

File: apps/soc-analyzer-pc/src-tauri/src/adb.rs
use std::process::Command;
use crate::DeviceInfo;

pub async fn scan_devices() -> Result<Vec<DeviceInfo>, Box<dyn std::error::Error>> {
    let output = Command::new("adb")
        .args(["devices", "-l"])
        .output()?;

    let output_str = String::from_utf8_lossy(&output.stdout);
    let mut devices = Vec::new();

    for line in output_str.lines().skip(1) {
        if line.trim().is_empty() {
            continue;
        }

        let parts: Vec<&str> = line.split_whitespace().collect();
        if parts.len() >= 2 && parts[1] == "device" {
            let device_id = parts[0].to_string();
            
            let model = get_device_property(&device_id, "ro.product.model").await
                .unwrap_or_else(|_| "Unknown".to_string());
            let brand = get_device_property(&device_id, "ro.product.brand").await
                .unwrap_or_else(|_| "Unknown".to_string());
            let soc = get_device_property(&device_id, "ro.board.platform").await
                .unwrap_or_else(|_| "Unknown".to_string());

            devices.push(DeviceInfo {
                id: device_id,
                name: format!("{} {}", brand, model),
                model,
                soc,
                connection_type: "usb".to_string(),
                status: "connected".to_string(),
            });
        }
    }

    Ok(devices)
}

async fn get_device_property(device_id: &str, prop: &str) -> Result<String, Box<dyn std::error::Error>> {
    let output = Command::new("adb")
        .args(["-s", device_id, "shell", "getprop", prop])
        .output()?;

    let value = String::from_utf8_lossy(&output.stdout).trim().to_string();
    Ok(value)
}

pub async fn forward_port(device_id: &str, local_port: u16, remote_port: u16) -> Result<(), Box<dyn std::error::Error>> {
    Command::new("adb")
        .args([
            "-s", device_id,
            "forward",
            &format!("tcp:{}", local_port),
            &format!("tcp:{}", remote_port)
        ])
        .output()?;
    
    Ok(())
}

pub async fn remove_forward(device_id: &str, local_port: u16) -> Result<(), Box<dyn std::error::Error>> {
    Command::new("adb")
        .args([
            "-s", device_id,
            "forward",
            "--remove",
            &format!("tcp:{}", local_port)
        ])
        .output()?;
    
    Ok(())
}

pub async fn push_file(device_id: &str, local_path: &str, remote_path: &str) -> Result<(), Box<dyn std::error::Error>> {
    Command::new("adb")
        .args(["-s", device_id, "push", local_path, remote_path])
        .output()?;
    
    Ok(())
}

pub async fn pull_file(device_id: &str, remote_path: &str, local_path: &str) -> Result<(), Box<dyn std::error::Error>> {
    Command::new("adb")
        .args(["-s", device_id, "pull", remote_path, local_path])
        .output()?;
    
    Ok(())
}
